File: application/backend-credit-card-enrollment/backend-c#/Common/SerializedEvent/Serializer.cs
using System.Text.Json.Nodes;
using CreditCardEnrollment.CreditCard.Enrollment.Event;

namespace CreditCardEnrollment.Common.SerializedEvent;

public class Serializer
{
    public SerializedEvent Serialize(Event.Event @event)
    {
        return new SerializedEvent
        {
            EventId = @event.EventId,
            AggregateId = @event.AggregateId,
            AggregateVersion = @event.AggregateVersion,
            CorrelationId = @event.CorrelationId,
            CausationId = @event.CausationId,
            RecordedOn = FormatDateTime(@event.RecordedOn),
            EventName = DetermineEventName(@event),
            JsonPayload = CreateJsonPayload(@event),
            JsonMetadata = "{}"
        };
    }

    private static string DetermineEventName(Event.Event @event)
    {
        return @event switch
        {
            EnrollmentRequested => "CreditCard_Enrollment_EnrollmentRequested",
            EnrollmentAccepted => "CreditCard_Enrollment_EnrollmentAccepted",
            EnrollmentDeclined => "CreditCard_Enrollment_EnrollmentDeclined",
            _ => throw new ArgumentException($"Unknown event type: {@event.GetType().Name}")
        };
    }

    private static string CreateJsonPayload(Event.Event @event)
    {
        var jsonObject = new JsonObject();

        switch (@event)
        {
            case EnrollmentRequested enrollmentRequested:
                jsonObject.Add("userId", enrollmentRequested.UserId);
                jsonObject.Add("productId", enrollmentRequested.ProductId);
                jsonObject.Add("annualIncomeInCents", enrollmentRequested.AnnualIncomeInCents);
                break;

            case EnrollmentAccepted enrollmentAccepted:
                jsonObject.Add("reasonCode", enrollmentAccepted.ReasonCode);
                jsonObject.Add("reasonDescription", enrollmentAccepted.ReasonDescription);
                break;

            case EnrollmentDeclined enrollmentDeclined:
                jsonObject.Add("reasonCode", enrollmentDeclined.ReasonCode);
                jsonObject.Add("reasonDescription", enrollmentDeclined.ReasonDescription);
                break;
        }

        return jsonObject.ToJsonString();
    }

    private static string FormatDateTime(DateTime dateTime)
    {
        return dateTime.ToUniversalTime().ToString("yyyy-MM-dd HH:mm:ss.ffffff UTC");
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/common/serializedEvent/Serializer.ts
import { Event } from '../event/Event';
import { SerializedEvent } from './SerializedEvent';
import {EnrollmentRequested} from "../../creditCard/enrollment/event/EnrollmentRequested";
import {EnrollmentAccepted} from "../../creditCard/enrollment/event/EnrollmentAccepted";
import {EnrollmentDeclined} from "../../creditCard/enrollment/event/EnrollmentDeclined";
import {injectable} from "tsyringe";

@injectable()
export class Serializer {
    serialize(event: Event): SerializedEvent {
        return {
            eventId: event.eventId,
            aggregateId: event.aggregateId,
            aggregateVersion: event.aggregateVersion,
            correlationId: event.correlationId,
            causationId: event.causationId,
            recordedOn: this.formatDateTime(event.recordedOn),
            eventName: this.determineEventName(event),
            jsonPayload: this.createJsonPayload(event),
            jsonMetadata: '{}'
        };
    }

    private determineEventName(event: Event): string {
        if (event instanceof EnrollmentRequested) {
            return 'CreditCard_Enrollment_EnrollmentRequested';
        }
        if (event instanceof EnrollmentAccepted) {
            return 'CreditCard_Enrollment_EnrollmentAccepted';
        }
        if (event instanceof EnrollmentDeclined) {
            return 'CreditCard_Enrollment_EnrollmentDeclined';
        }
        throw new Error(`Unknown event type: ${event.constructor.name}`);
    }

    private createJsonPayload(event: Event): string {
        const payload: Record<string, any> = {};

        if (event instanceof EnrollmentRequested) {
            payload.userId = event.userId;
            payload.productId = event.productId;
            payload.annualIncomeInCents = event.annualIncomeInCents;
        } else if (event instanceof EnrollmentAccepted) {
            payload.reasonCode = event.reasonCode;
            payload.reasonDescription = event.reasonDescription;
        } else if (event instanceof EnrollmentDeclined) {
            payload.reasonCode = event.reasonCode;
            payload.reasonDescription = event.reasonDescription;
        }

        return JSON.stringify(payload);
    }

    private formatDateTime(date: Date): string {
        return date.toISOString().replace('Z', ' UTC');
    }
}